http/events.go

```diff
@@ -69,14 +87,80 @@ func handleEventsSubscribeWebsocket(args eventSubscribeArgs) (websocket.StatusCo
 				logger.Warn("Could not serialize websocket event", "error", err)
 				return 0, "", err
 			}
-			err = args.conn.Write(ctx, messageType, messageBytes)
+			err = sub.conn.Write(ctx, messageType, messageBytes)
 			if err != nil {
 				return 0, "", err
 			}
 		}
 	}
 }
 
+// allowMessageCached checks that the message is allowed to received by the websocket.
+// It caches results for specific namespaces, data paths, and event types.
+func (sub *eventSubscriber) allowMessageCached(message *logical.EventReceived) bool {
+	// default to the subscribe endpoint if there is no specific data path
+	messageNs := strings.Trim(message.Namespace, "/")
+	dataPath := "sys/events/subscribe"
+	if message.Event.Metadata != nil {
+		dataPathField := message.Event.Metadata.GetFields()[logical.EventMetadataDataPath]
+		if dataPathField != nil {
+			dataPath = dataPathField.GetStringValue()
+		}
+	}
+	cacheKey := fmt.Sprintf("%v!%v!%v", messageNs, dataPath, message.EventType)
+	_, ok := sub.checkCache.Get(cacheKey)
+	if ok {
+		return true
+	}
+
+	// perform the actual check and cache it if true
+	ok = sub.allowMessage(messageNs, dataPath, message.EventType)
+	if ok {
+		err := sub.checkCache.Add(cacheKey, ok, webSocketRevalidationTime)
+		if err != nil {
+			sub.logger.Debug("Error adding to policy check cache for websocket", "error", err)
+			// still return the right value, but we can't guarantee it was cached
+		}
+	}
+	return ok
+}
+
+// allowMessage checks that the message is allowed to received by the websocket
+func (sub *eventSubscriber) allowMessage(eventNs, dataPath, eventType string) bool {
+	// does this even match the requested namespaces
+	matchedNs := false
+	for _, nsPattern := range sub.namespacePatterns {
+		if glob.Glob(nsPattern, eventNs) {
+			matchedNs = true
+			break
+		}
+	}
+	if !matchedNs {
+		return false
+	}
+
+	// next check for specific access to the namespace and event types
+	nsDataPath := dataPath
+	if eventNs != "" {
+		nsDataPath = path.Join(eventNs, dataPath)
+	}
+	capabilities, eventTypes, err := sub.core.CapabilitiesAndSubscribeEventTypes(sub.ctx, sub.clientToken, nsDataPath)
+	if err != nil {
+		sub.logger.Debug("Error checking capabilities and event types for token", "error", err, "namespace", eventNs)
+		return false
+	}
+	if !(slices.Contains(capabilities, "root") || slices.Contains(capabilities, "subscribe")) {
+		return false
+	}
+	for _, pattern := range eventTypes {
```

Comment:
nit: Maybe just me but I found the loop quite confusing to read because of how similar the `eventType` and `eventTypes` variable names are without them being from the same data source.

```suggestion
	capabilities, allowedEventTypes, err := sub.core.CapabilitiesAndSubscribeEventTypes(sub.ctx, sub.clientToken, nsDataPath)
	if err != nil {
		sub.logger.Debug("Error checking capabilities and event types for token", "error", err, "namespace", eventNs)
		return false
	}
	if !(slices.Contains(capabilities, "root") || slices.Contains(capabilities, "subscribe")) {
		return false
	}
	for _, pattern := range allowedEventTypes {
```